lib: Introduce generic (page guarded) stack allocator library

mogasergiu opened the pull request:
<!--

Thank you for opening a new PR to the Unikraft Open Source Project!  We welcome
new changes, features, fixes, and more!  Please fill in this form to indicate
the status of your PR.  Please ensure you have read the contribution guidelines
before opening a new PR as this will cover the PR process:

  https://unikraft.org/docs/contributing/

-->

### Prerequisite checklist

<!--
Please mark items appropriately:
-->

 - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
 - [x] Tested your changes against relevant architectures and platforms;
 - [x] Ran the [`checkpatch.uk`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.uk) on your commit series before opening this PR;
 - [ ] Updated relevant documentation.


### Base target

 - Architecture(s): N/A
 - Platform(s): N/A
 - Application(s): N/A


### Additional configuration

<!--
Please specify any additional configuration which is needed for this feature to
work or any new configuration parameters which are introduced by this PR.  This
will help during the review process.  For example:

 - `CONFIG_LIBUKDEBUG=y`

-->

### Description of changes

<!--
Please provide a detailed description of the changes made in this new PR.
-->

Implement a generic stack allocation wrapper over a given Unikraft
memory allocator. The library receives as arguments an allocator and,
if libukvmem is enabled, a virtual address space and an initial size.
The library works in three ways:
1. If libukvmem is disabled, the library will make use of the allocator
received as argument to allocate the initial allocator structure and
Unikraft compliant stacks.
2. If libukvmem is enabled, the library will make use of the allocator
received as argument to allocate the initial allocator structure and
Unikraft compliant stacks. The difference from 1. is that the virtual
address space and the initial size arguments are now available. The
two arguments will hint how much of the stack to premap whenever they
are allocated first, i.e. what is the initial stack size before
on-demand paging kicks in.
3. If libukvmem is enabled AND the configuration option of the library
integrating page guards is enabled, the allocator given as argument
will only be used once when allocating the initial allocator structure,
while the virtual address space and initial size arguments will be used
to create stack VMA's, specific to libukvmem.

With this addition, now libukboot is made aware of this stack allocator
and therefore a new configuration option is introduced to allow the
customization of the aforementioned initial size argument for when
libukvmem is enabled. Consequently, libukschedcoop's initialization
allows now for three additional allocator arguments for regular
stacks, for auxiliary stacks and for the TLS. This is done so that
libukboot can now initialize and pass to libukschedcoop separate
stack allocators, differentiated by the differently configured size,
alignment as well as the fact that, unlike regular stacks,
the auxiiliary stacks must always be fully mapped.

Also a quick basic commit sprinkling some error messages in the frame allocator :man_shrugging:.

mogasergiu committed:
{arch,lib,plat}: Align auxstack with regular stack configuration

Everything stack size or alignment related is defined in `arch/`.
Therefore, move everything auxiliary stack related there and rename
related definitions/configurations to have them aligned with the
normal stacks. Additionally, remove the no longer useful related
method `ukplat_auxsp_alloc`.

Signed-off-by: Sergiu Moga <sergiu@unikraft.io>

mogasergiu committed:
lib: Introduce generic (page guarded) stack allocator library

Implement a generic stack allocation wrapper over a given Unikraft
memory allocator. The library receives as arguments an allocator and,
if libukvmem is enabled, a virtual address space and an initial size.
The library works in three ways:
1. If libukvmem is disabled, the library will make use of the allocator
received as argument to allocate the initial allocator structure and
Unikraft compliant stacks.
2. If libukvmem is enabled, the library will make use of the allocator
received as argument to allocate the initial allocator structure and
Unikraft compliant stacks. The difference from 1. is that the virtual
address space and the initial size arguments are now available. The
two arguments will hint how much of the stack to premap whenever they
are allocated first, i.e. what is the initial stack size before
on-demand paging kicks in.
3. If libukvmem is enabled AND the configuration option of the library
integrating page guards is enabled, the allocator given as argument
will only be used once when allocating the initial allocator structure,
while the virtual address space and initial size arguments will be used
to create stack VMA's, specific to libukvmem.

Signed-off-by: Sergiu Moga <sergiu@unikraft.io>

mogasergiu committed:
lib/ukvmem: Make stack VMA guards size configurable and end-to-end

Add end-to-end page guards to stack VMAs and make their size be
configurable: default 4 pages for top (to counter stack declared arrays)
and 1 page for bottom (to counter too many pop instructions).

Signed-off-by: Sergiu Moga <sergiu@unikraft.io>
Checkpatch-Ignore: LONG_LINE_COMMENT

mogasergiu commented on lib/ukallocstack/stack.c:
```diff
@@ -0,0 +1,277 @@
+	return addr;
+}
+
+struct uk_alloc *uk_allocstack_init(struct uk_alloc *a,
```
Missed a spot, no trailing comma here.

mogasergiu commented on lib/ukallocstack/stack.c:
```diff
@@ -0,0 +1,277 @@
+	UK_ASSERT(a);
+#if CONFIG_LIBUKVMEM
+	UK_ASSERT(vas);
+#endif /* CONFIG_LIBUKALLOCSTACK_PAGE_GUARDS */
```
Note: just noticed I have lots of `CONFIG_LIBUKALLOCSTACK_PAGE_GUARDS` instead of `CONFIG_LIBUKVMEM` comments.

mogasergiu commented on lib/ukallocstack/stack.c:
```diff
@@ -0,0 +1,277 @@
+#include <uk/vmem.h>
+#endif /* CONFIG_LIBUKVMEM */
+
+struct uk_stackalloc {
```
TODO: s/stackalloc/allocstack

michpappas commented on lib/ukallocstack/include/uk/allocstack.h:
```diff
@@ -0,0 +1,27 @@
+extern "C" {
+#endif
+
+struct uk_alloc *uk_allocstack_init(struct uk_alloc *a
```
As this is the public API, please consider documenting to avoid confusion (e.g. what is the purpose of `initial_size`), incl. all provided ops and the constraints in their parameters. This is especially important for allocator-specific variations from the usual constraints, such as the alignment restrictions in `[posix_]memalign()`.

michpappas commented on lib/ukallocstack/include/uk/allocstack.h:
```diff
@@ -0,0 +1,27 @@
+#if CONFIG_LIBUKVMEM
+				     , struct uk_vas __maybe_unused *vas,
+				     __sz initial_size
+#endif /* CONFIG_LIBUKALLOCSTACK_PAGE_GUARDS */
```
Should be `#endif /* CONFIG_LIBUKVMEM */`

michpappas commented on lib/ukallocstack/stack.c:
```diff
@@ -0,0 +1,277 @@
+#if CONFIG_LIBUKVMEM
+	sa->vas = vas;
+	sa->initial_size = initial_size;
+#endif /* !CONFIG_LIBUKVMEM */
```
Should be `#endif /* CONFIG_LIBUKVMEM */`.

michpappas commented on lib/ukallocstack/stack.c:
```diff
@@ -0,0 +1,277 @@
+
+	sa = to_allocstack(a);
+
+	uk_free(sa->og_a, sa);
```
~~Also free(sa)~~ Sorry I didn't read this correctly.

michpappas commented on lib/ukallocstack/stack.c:
```diff
@@ -0,0 +1,277 @@
+
+	UK_ASSERT(a);
+
+	if (!size)
```
`unlikely`?

mogasergiu replied:
Regarding this and the other `unlikely` comments. I meant to align with the default `memalign`/`posix-memalign` checks which don't have these. But now, looking back, it's better to have them here in this case, since this lib is only meant to be used by our own kernel internal allocators :+1:

michpappas commented on lib/ukallocstack/stack.c:
```diff
@@ -0,0 +1,277 @@
+#endif /* CONFIG_LIBUKVMEM */
+
+struct uk_allocstack {
+	struct uk_alloc a;
```
I would had called these `sa->sa` and `sa->a` instead of `sa->a` and `sa->og_a` mostly because I don't like the `og_a` vary much, but maybe that's a matter of personal taste, so feel free to ignore if you prefer the current convention.

mogasergiu replied:
I think I like `sa->a` and `sa->parent_a` :+1:

michpappas commented on lib/ukallocstack/stack.c:
```diff
@@ -0,0 +1,277 @@
+	struct uk_allocstack *sa;
+	int rc __maybe_unused;
+	__vaddr_t vaddr;
+	__sz real_size;
```
Heheh, at least you didn't use [`desired_size`](https://github.com/unikraft/unikraft/pull/1212/files#diff-628d5113b18ee6253ddbcc848991834d8493a6ad8684eea4e944be9252f73340L79) 🤤 this time. `alloc_size` could be a more elegant name.

michpappas commented on lib/ukallocstack/stack.c:
```diff
@@ -0,0 +1,277 @@
+	sa = to_allocstack(a);
+
+#if CONFIG_LIBUKALLOCSTACK_PAGE_GUARDS
+	real_size = PAGE_ALIGN_UP(size) + PAGE_SIZE;
```
Perhaps it would be useful to add a comment about the purpose of that extra page.

mogasergiu replied:
hmm i think it's actually not needed anymore as I [removed](https://github.com/unikraft/unikraft/pull/1322/commits/3bf2bc5de8824ae55bd812ec011d9be27914a2c6#diff-af0f86d6a09f05c4a44e859d58dfe38ab95d442f01f60e26f92b345aae167f8fR124) this constraint.

michpappas commented on lib/ukallocstack/Config.uk:
```diff
@@ -0,0 +1,22 @@
+menuconfig LIBUKALLOCSTACK
+	bool "ukallocstack: Generic stack allocator wrapper"
```
Not sure if the "wrapper" part is accurate here, given that if guards are enabled it is no longer just a wrapper. See also the comment below.

michpappas commented on lib/ukallocstack/Config.uk:
```diff
@@ -0,0 +1,22 @@
+	select LIBUKDEBUG
+	select LIBUKALLOC
+	help
+		This is a generic stack allocating library that is registered
```
So the main value of this library is the guards. And being able to use the library without vmem is just a convenience that allows the user to make a single call without conditionals.

Given that, I would probably reverse the logic and state here that without stack guards enabled this acts as a simple wrapper around the allocator. Similarly, in the description of the guards, I wouldn't describe it as an exception of the wrapper case (i.e. "instead of being a wrapper").

PS: There is also a school of thought that says that you should not describe the implementation at all, but only what the feature provides.

michpappas commented on lib/ukallocstack/stack.c:
```diff
@@ -0,0 +1,277 @@
+	}
+#if CONFIG_LIBUKVMEM
+	sa->vas = vas;
+	sa->initial_size = initial_size;
```
You may consider making it a requirement that the user passes an aligned `initial_size`, so that you don't need to align up in every individual op.

michpappas commented on lib/ukallocstack/stack.c:
```diff
@@ -0,0 +1,277 @@
+
+#if CONFIG_LIBUKVMEM
+	sp = ukarch_gen_sp(vaddr, real_size);
+	rc = uk_vma_advise(sa->vas,
```
You may want to take into account the possibility that `size` < `initial_size`, or require that the user passes `size` >= `initial_size`.

mogasergiu replied:
It's fine actually, that would mean the whole stack would be paged-in. I will document this.

michpappas replied:
What I actually meant by `stack_size < initial_size` was under that condition we ask madvise to page in more memory than the actual stack. Therefore we need a check somewhere that `initial_size <= size`.

michpappas commented on lib/ukallocstack/stack.c:
```diff
@@ -0,0 +1,277 @@
+			  vaddr,
+			  vma->end - vma->start,
+			  0);
+	UK_ASSERT(!rc);
```
Same here, `UK_BUGON(rc)` if you choose to adopt the previous suggestion.

michpappas commented on lib/ukallocstack/stack.c:
```diff
@@ -0,0 +1,277 @@
+	/* Crash if pointer given to free is not part of any VMA. This would
+	 * normally lead to a heap corruption.
+	 */
+	UK_ASSERT(vma);
```
Maybe it's a matter of personal taste but as far as semantics go `UK_BUGON(!vma)` would be more suitable here.

michpappas commented on lib/ukallocstack/stack.c:
```diff
@@ -0,0 +1,277 @@
+
+	UK_ASSERT(a);
+
+	if (nmemb <= 1 || !size)
```
`unlikely`?

michpappas commented on lib/ukallocstack/stack.c:
```diff
@@ -0,0 +1,277 @@
+	UK_ASSERT(align >= UKARCH_SP_ALIGN);
+	UK_ASSERT(IS_ALIGNED(align,  UKARCH_SP_ALIGN));
+
+	if (!size)
```
`unlikely`?

michpappas commented on lib/ukallocstack/stack.c:
```diff
@@ -0,0 +1,277 @@
+	UK_ASSERT(memptr);
+	UK_ASSERT(align >= UKARCH_SP_ALIGN);
+	UK_ASSERT(IS_ALIGNED(align,  UKARCH_SP_ALIGN));
+#if CONFIG_LIBUKALLOCSTACK_PAGE_GUARDS
```
So on the one hand you make this check here, but on the other hand you delegate the `size` check to `stack_memalign()`. Maybe choose one scheme to be consistent?

michpappas commented on lib/ukallocstack/stack.c:
```diff
@@ -0,0 +1,277 @@
+
+	real_size = PAGE_ALIGN_UP(size) + PAGE_SIZE;
+	vaddr = __VADDR_ANY;
+	rc = uk_vma_map_stack(sa->vas,
```
Shouldn't you adjust the return vaddr to the requested alignment?

mogasergiu replied:
If using page guards then we allocate stack VMA's instead. So, the return address will always be page-aligned and the maximum allowed alignment is PAGE_SIZE because we cannot satisfy higher alignments and smaller than PAGE_SIZE alignments must be a divisor of PAGE_SIZE.

I wil add a comment

mogasergiu replied:
The only way we could do that is, I guess, if we make the following adjustment, and instead allocating bigger VMA's of size `alloc_size + align` and returning the `vaddr` as `vaddr = ALIGN_UP(vaddr, align)`, but at that point it becomes wasteful I would say as this is a specialized allocator we use in very specific cases, hence why I agreed to the `unlikely`'s as well (I also like them cus they make me look smart). I will add a comment in case this will be needed in the future to know what to do, but I highly doubt it. (yes i am also too lazy to make the change now :cowboy_hat_face: )

michpappas replied:
Okay since we have a fundamental limitation on alignment by guards, perhaps we should at least document this constraint, given that it deviates from the expected behavior, and also add a check.

michpappas reviewed (changes requested):
@mogasergiu nice 😎  Some review comments below.

michpappas commented on arch/Config.uk:
```diff
@@ -44,6 +44,15 @@ config CPU_EXCEPT_STACK_SIZE_PAGE_ORDER
 		2 ^ order * page size (e.g. 4KB).
 		Only change this if you know what you're doing.
 
+config AUXSTACK_SIZE_PAGE_ORDER
```
I am posting this here as GitHub does not provide a way to comment on the commit message: 

`STACK_SIZE` should not had been under `arch/` in the first place. As discussed before, we should be moving everything not defined by the architecture away from `arch/`. Therefore this change does not improve something, both the current and the new place will be changed again shortly.

An additional problem is that this change propagates to an additional change in Kconfig. In principle Kconfig directives is something we should treat as a stable ABI and limit changes on them to the absolutely necessary, as such changes have an impact both on the catalog (they must be propagated to Kraftfiles) and potentially to systems maintained by adopters of Unikraft - that's even worse as we have no control over these. Of course under re-arch these are expected to change a lot, but still we should avoid them when they are not necessary.

Moreover, various stack sizes it makes more sense for the various stack sizes to be defined by the platform rather than the architecture. My take is that they should live in the upcoming `platform.h`. So I would propose that we drop this change, and move them to `platform.h` as soon as we have consensus on that file with the rest of the maintainers (I'll bring it up on the next call). Alternatively, instead of dropping - and assuming there will be consensus on `platform.h` - we can defer merging and update this commit to move them to that place. What do you think?

michpappas replied:
**Update:** During an offline discussion with @skuenzer it was agreed that we merge this as is, given that migration involves additional code that existed previously in `arch`, and do a migration once all depending PRs are merged.

michpappas commented on lib/ukvmem/Config.uk:
```diff
@@ -71,6 +71,18 @@ config LIBUKVMEM_STACK_BASE
 
+config LIBUKVMEM_STACK_GUARD_PAGES_TOP
+	int "Top guard pages count"
+	default 4
```
Although these are not costly, similarly to timeouts there are no good default values. So I would be curious to know where the selected default come from, are 4 pages found to be good enough for most cases? Also, how common are accesses beyond 1 page?

mogasergiu replied:
It's based on intuition after noticing that 2 pages are not enough for a Wordpress based application.

michpappas replied:
Random application (e.g. worpress) sounds like an arbitrary criterion though. Similarly to timeouts, there's no such thing as a good default, as there will always either too much or too little. Personally I would choose 1 and let the user adjust to the application requirements. Unless we have strong evidence of an average.

@skuenzer what are your thoughts on this?

skuenzer replied:
In fact the issue was related to on-stack array access. Since stacks are growing downwards and arrays upwards, you easily get outside of a single 4k page when accessing index 0 of a bigger array. On one hand I agree with you @michpappas and on the other hand this should be a protective feature. I found that there is a [POSIX interface](https://pubs.opengroup.org/onlinepubs/009695399/functions/pthread_attr_getguardsize.html) for it, so it should be ultimately a per-thread property to make it right. I can imagine that the rational behind this coulb probably be that an application can estimate typical access distances and size the guard pages accordingly.
At this point, I would accept the 4 pages for user experience but longer term we need to change the integration so that it becomes configurable per thread -> I think this requires some API discussion first.

michpappas commented on lib/ukvmem/include/uk/vma_types.h:
```diff
@@ -109,32 +109,41 @@ extern const struct uk_vma_ops uk_vma_stack_ops;
 	__vaddr_t va;
+	int rc;
 
-	UK_ASSERT(len >= 2 * PAGE_SIZE); /* one page + guard page */
```
Same here. If there is a requirement for increasing the number of guard pages to >1, then it would be worth providing the reasoning in the commit message.

Moreover, should any of the tests that involve guards need to be updated? I did a quick run and they all pass, but didn't look at the implementation to determine whether they are still valid or they just produce false positives.

mogasergiu replied:
> Moreover, should any of the tests that involve guards need to be updated?

They are still fine, the new guard pages also include the old guard page.

michpappas replied:
Yes, but that sounds that they are now incomplete as they only test one guard page?

michpappas commented on lib/ukvmem/Config.uk:
```diff
@@ -71,6 +71,18 @@ config LIBUKVMEM_STACK_BASE
+
+config LIBUKVMEM_STACK_GUARD_PAGES_BOTTOM
+	int "Bottom guard pages count"
+	default 4
```
For the stack base, is it meaningful to have more than 1 page? The only case I can think of is a case where a bug (benign or not) causes an overwrite beyond the stack base.

mogasergiu replied:
> The only case I can think of is a case where a bug (benign or not) causes an overwrite beyond the stack base.

Yep, following a discussion with @skuenzer it was decided to make the guard pages end-to-end.

michpappas replied:
Okay, but what is the rationale for that? @skuenzer

michpappas commented on lib/ukvmem/include/uk/vma_types.h:
```diff
@@ -109,32 +109,41 @@ extern const struct uk_vma_ops uk_vma_stack_ops;
 {
-	int rc;
 	__vaddr_t va;
+	int rc;
```
I see what you did there 🎄

mogasergiu replied:
World domination next :christmas_tree: :santa:

michpappas reviewed (changes requested):
@mogasergiu a few more

michpappas commented on lib/ukallocstack/Config.uk:
"Useful"

michpappas commented on lib/ukallocstack/include/uk/allocstack.h:
```diff
@@ -0,0 +1,71 @@
+ *
+ * NOTE: When using the page guards configuration, the maximum allowed
+ * alignment for functions such as memalign/posix_memalign is PAGE_SIZE.
+ * See comment from `stack_memalign()`.
```
Nit: "See comment in". 

Also consider whether it would be useful to additionally remind the reader that this is in addition to the usual requirement that alignment must be a power of 2.

michpappas commented on lib/ukallocstack/include/uk/allocstack.h:
```diff
@@ -0,0 +1,71 @@
+ *     them
+ *     - initial_size == 0 => no page of any of the allocated stacks is
+ *     pre-mapped
+ *     - initial_size == SOME_STACK_SIZE then a uk_malloc(SOME_STACK_SIZE) would
```
I don't know why `SOME` initially confused me, `STACK_SIZE` would had been immediately clear to me. But this may be subjective, so if you think it's good as is please feel free to ignore.

mogasergiu replied:
Will surround it with `< >` to further suggest it's just some placeholder.

michpappas reviewed (changes requested):
@mogasergiu a few comments on the last update.

michpappas commented on lib/ukschedcoop/include/uk/schedcoop.h:
```diff
@@ -44,7 +44,10 @@
 #endif
 
-struct uk_sched *uk_schedcoop_create(struct uk_alloc *a);
+struct uk_sched *uk_schedcoop_create(struct uk_alloc *a,
```
Leaving this as a note for @skuenzer. Not sure if it's something to do in the scope of this PR, but perhaps it would be worth to consider passing a struct with allocators similarly to `ops`. Throughout these PRs I have noticed that many functions in `thread.h` now take 10 or more parameters.

skuenzer replied:
I understand what you mean. For now, the creation functions for scheduler initialization are custom and defined by the scheduler implementation, likewise the allocator initialization. There is no common init interface layout defined yet for schedulers and allocators. I am also not sure if we could define one because schedulers and allocators could handle very specific cases with very custom parameters.

skuenzer replied:
Concluding: I think it is okay to expand this definition with this PR. The only place where the scheduler is initialized is in ukboot atm, impact should be minimal for this particular case.

michpappas reviewed (changes requested):
From my side it's good. Remaining work before RB:
* Update tests / justify that they don't need to be updated.
* Open questions that require input from @skuenzer.
*  Break down into per-library commits before we merge.

@skuenzer this one's on you now.

mogasergiu committed:
lib/ukvmem: Fix out of range `vma_end` found as last element of list

Consider the following case:
               +------------+
               |VMA to unmap|
               +------------+      next VMA
          +----------+           +----------+
          |mapped VMA|           |mapped VMA|
          +----------+           +----------+
The current state of the code would simply assign `vma_end` to the
next mapped VMA without checking if the end of the VMA to unmap is
within its range. Fix this by adding this check and ensure we do
not assign `vma_end` to the next VMA if it does not meet the
requirements.

Signed-off-by: Sergiu Moga <sergiu@unikraft.io>

mogasergiu commented:
TODO: Split last commit into three
- introduce the lib
- extend schedcoop function for the other three allocators
- adjust ukboot to use stack allocator

mogasergiu commented:
TODO: Following a discussion it's been decided:
- use only 1 guard page for bottom guards
- instead of `initial_len`/`initial_size` use `premapped_len`
- remove the termination function: "Since we generally cannot do it right, do not do it for now"

mogasergiu committed:
lib/uksched: Give scheduler init higher granularity w.r.t. allocators

Instead of using the same allocator for stack, auxiliary stack, TLS
and other structures, allow a caller of the scheduler initialization
function to pass different allocators for these if they want to.

Signed-off-by: Sergiu Moga <sergiu@unikraft.io>

mogasergiu committed:
lib/ukboot: Make use of the generic (page guarded) stack allocator

Allow ukboot, if configured to, to make use of the stack allocator
library for higher flexibilty and more control over the allocators used
for stacks. Specifically, initialize the scheduler with special stack
allocators and, because of this, remove the explicit `uk_vma_advise`
from uksched's thread creation and allow these allocators to
transparently do what they were configured to.

The default initialization of the cooperative scheduler now takes a
normal stack allocator for generic thread stacks whose initial
premapped length is configurable through the
CONFIG_LIBUKBOOT_ALLOCSTACK_PREMAP_ORDER configuration entry. What is
more, we keep the auxiliary stack semantics of always having it
premapped by configuring its stack allocator to fully map any stack
that is generated through it.

Signed-off-by: Sergiu Moga <sergiu@unikraft.io>

michpappas reviewed (approved):
Reviewed-by: Michalis Pappas <michalis@unikraft.io>

skuenzer commented on lib/ukboot/boot.c:
```diff
@@ -64,6 +65,11 @@
 #include <uk/tinyalloc.h>
 #define uk_alloc_init uk_tinyalloc_init
 #endif
+#include <uk/allocstack.h>
```
```c
#if CONFIG_LIBUKBOOT_ALLOCSTACK
```
etc. is missing.

skuenzer replied:
The following patch makes ukboot still compile even without allocator initialzation:
```diff
--- a/lib/ukboot/boot.c
+++ b/lib/ukboot/boot.c
@@ -65,11 +65,13 @@
 #include <uk/tinyalloc.h>
 #define uk_alloc_init uk_tinyalloc_init
 #endif
+#if CONFIG_LIBUKBOOT_ALLOCSTACK
 #include <uk/allocstack.h>
 #if CONFIG_LIBUKBOOT_ALLOCSTACK_PREMAP_ORDER
 #define ALLOCSTACK_INITIAL_SIZE                                \
        (PAGE_SIZE * (1 << CONFIG_LIBUKBOOT_ALLOCSTACK_PREMAP_ORDER))
 #endif /* CONFIG_LIBUKBOOT_ALLOCSTACK_PREMAP_ORDER */
+#endif /* CONFIG_LIBUKBOOT_ALLOCSTACK */
 #if CONFIG_LIBUKSCHED
 #include <uk/sched.h>
 #endif /* CONFIG_LIBUKSCHED */
```

skuenzer commented on include/uk/plat/config.h:
```diff
@@ -42,5 +42,8 @@
 
+#define AUXSTACK_SIZE_PAGE_ORDER		\
+	__AUXSTACK_SIZE_PAGE_ORDER
+#define AUXSTACK_SIZE				__AUXSTACK_SIZE
```
With re-arch we definitely have to fix the namespacing of the definitions in this file. For now, okay since the newly introduced parameters follows the given examples.

skuenzer commented on lib/ukvmem/include/uk/vma_types.h:
```diff
@@ -86,66 +86,83 @@ static inline int uk_vma_map_anon(struct uk_vas *vas, __vaddr_t *vaddr,
 /* Stack flags */
 #define UK_VMA_STACK_GROWS_UP		(0x1UL << UK_VMA_MAP_EXTF_SHIFT)
 
+#define STACK_TOP_GUARD_SIZE				\
```
Could we at least namespace those, like `UK_VMA_STACK_TOP_GUARD_SIZE`? We have previous definitions in this file already namespaced.

skuenzer commented on lib/ukallocstack/include/uk/allocstack.h:
```diff
@@ -0,0 +1,63 @@
+ */
+struct uk_alloc *uk_allocstack_init(struct uk_alloc *a
+#if CONFIG_LIBUKVMEM
+				     , struct uk_vas __maybe_unused *vas,
```
Can this `maybe_unused` really happen? With `CONFIG_LIBUKVMEM` on, it should be used always, right? I would also prefer if we add this attribute to the final function declaration instead.

skuenzer reviewed (approved):
Thanks a lot for this work!

Approved-by: Simon Kuenzer <simon@unikraft.io>
